src/state/slices/opportunitiesSlice/resolvers/idle.ts

```diff
@@ -0,0 +1,100 @@
+import type { AssetId } from '@shapeshiftoss/caip'
+import { fromAssetId, toAssetId } from '@shapeshiftoss/caip'
+import { DefiProvider, DefiType } from 'features/defi/contexts/DefiManagerProvider/DefiCommon'
+import { getIdleInvestor } from 'features/defi/contexts/IdleProvider/idleInvestorSingleton'
+import { logger } from 'lib/logger'
+import { selectAssetById } from 'state/slices/selectors'
+
+import type {
+  GetOpportunityIdsOutput,
+  GetOpportunityMetadataOutput,
+  OpportunitiesState,
+  OpportunityDefiType,
+  StakingId,
+} from '../types'
+import type { ReduxApi } from './types'
+
+const moduleLogger = logger.child({ namespace: ['opportunities', 'resolvers', 'idle'] })
+
+export const idleStakingOpportunitiesMetadataResolver = async ({
+  opportunityType,
+  reduxApi,
+}: {
+  opportunityType: OpportunityDefiType
+  reduxApi: ReduxApi
+}): Promise<{ data: GetOpportunityMetadataOutput }> => {
+  const idleInvestor = getIdleInvestor()
+  const opportunities = await idleInvestor.findAll()
+
+  const { getState } = reduxApi
+  const state: any = getState() // ReduxState causes circular dependency
+
+  const byId = await opportunities.reduce(
+    async (acc: Promise<OpportunitiesState[DefiType.Staking]['byId'] | undefined>, opportunity) => {
+      const assetId = toAssetId({
+        assetNamespace: 'erc20',
+        assetReference: opportunity.id,
+        chainId: fromAssetId(opportunity.feeAsset.assetId).chainId,
+      })
+
+      let accResolved = await acc
+      if (!accResolved) accResolved = {}
+
+      const asset = selectAssetById(state, assetId)
+
+      // Asset doesn't exist in portfolio, meaning this asset is bogus, e.g these two
+      // https://etherscan.io/address/0xa0154a44c1c45bd007743fa622fd0da4f6d67d57
+      // https://etherscan.io/address/0x5f45a578491a23ac5aee218e2d405347a0fafa8e
+      if (!asset) return accResolved
+
+      const rewardAssetIds = (await opportunity.getRewardAssetIds().catch(error => {
+        moduleLogger.debug(
+          { fn: 'idleStakingOpportunitiesMetadataResolver', error },
+          'Error fetching Idle opportunities metadata',
+        )
+      })) as [AssetId] | [AssetId, AssetId] | [AssetId, AssetId, AssetId] | undefined
```

Comment:
Classic TS issue, similar to module definitions defining 10+ lines of the same fn with diff arities. We might want to bring this guy over? https://stackoverflow.com/a/69248547/5927039

Reply:
Given there is really only 3 flavors of that tuple-length and I don't see us having different tuple lengths than 1/2/3 in the future, I'd rather keep the tuple of length, it effectively makes types act as documentation

WRT the alternative syntax, that would be breaking the types unfortunately:

```shell
    Type '[AssetId, (AssetId | undefined)?, (AssetId | undefined)?] | undefined' is not assignable to type 'readonly [AssetId, AssetId] | readonly [AssetId] | readonly [AssetId, AssetId, AssetId] | undefined'.
```